0012-common-ledger-api/0012-common-ledger-api.md

```diff
@@ -0,0 +1,1201 @@
+| ...                 | (Various)          | _(Optional)_ Additional arbitrary values as desired. |
+
+#### Example
+
+Request:
+
+```
+GET /
+```
+
+Response:
+
+```json
+HTTP/1.1 200 OK
+
+{
+  "currency_code": "EUR",
+  "currency_symbol": "€",
+  "connectors": [{
+    "id": "https://red.ilpdemo.org/ledger/accounts/connie",
+    "name": "connie"
+  }],
+  "urls": {
+    "health": "http://usd-ledger.example/health",
+    "transfers": "http://usd-ledger.example/transfers",
+    "transfer_fulfillment": "http://usd-ledger.example/transfers/{client_id}/fulfillment",
+    "account": "http://usd-ledger.example/accounts/{name}",
+    "websocket": "wss://usd-ledger.example/websocket/"
+  },
+  "precision": 10,
+  "scale": 2
+}
+```
+
+
+#### Errors
+
+- This method does not return any errors under normal circumstances.
+
+
+### Prepare Transfer
+[Prepare Transfer]: #prepare-transfer
+
+Prepares a new transfer (conditional or unconditional) in the ledger. When a transfer becomes prepared, it executes immediately if there is no condition. If you specify `execution_condition`, the funds are held until the beneficiary [submits the matching fulfillment][Submit Fulfillment] or the `expires_at` time is reached.
+
+**Authorization:** The owner of an account MUST be able to prepare a conditional transfer that debits the account. Non-administrators MUST NOT be able to prepare a transfer that debits a different account. A ledger MAY disallow non-administrators from preparing unconditional transfers.
+
+#### Request Format
+
+POST a JSON [Transfer resource][] to the `transfers` [URL][].
+
+**Metadata URL Name:** `transfers`
+
+##### Body Parameters
+
+The message body should be a JSON [Transfer resource][].
+
+#### Response Format
+
+A successful result uses the HTTP response code **201 Created** and contains a JSON body with the [Transfer resource][] as saved.
+
+#### Example
+
+Request:
+
+```json
+POST /transfers
+Content-Type: application/json
+
+{
+  "client_id": "3a2a1d9e-8640-4d2d-b06c-84f2cd613204",
+  "ledger": "http://usd-ledger.example",
+  "debit_account": "http://usd-ledger.example/accounts/alice",
+  "credit_account": "http://usd-ledger.example/accounts/bob",
+  "amount": "50",
+  "execution_condition": "cc:0:3:8ZdpKBDUV-KX_OnFZTsCWB_5mlCFI3DynX5f5H2dN-Y:2",
+  "expires_at": "2016-10-25T08:41:59.795Z"
+}
+```
+
+Response:
+
+```json
+HTTP/1.1 201 Created
+
+{
+  "id": "http://usd-ledger.example/transfers/3a2a1d9e-8640-4d2d-b06c-84f2cd613204",
+  "client_id": "3a2a1d9e-8640-4d2d-b06c-84f2cd613204",
+  "ledger": "http://usd-ledger.example",
+  "debit_account": "http://usd-ledger.example/accounts/alice",
+  "credit_account": "http://usd-ledger.example/accounts/bob",
+  "amount": "50",
+  "execution_condition": "cc:0:3:8ZdpKBDUV-KX_OnFZTsCWB_5mlCFI3DynX5f5H2dN-Y:2",
+  "expires_at": "2016-10-25T08:41:59.795Z",
+  "state": "prepared",
+  "timeline": {
+    "prepared_at": "2015-06-16T00:00:01.000Z"
```

Comment:
Should the API maintain both values, allowing for a clear distinction between `created_at` and a `prepared_at`?  

Per RFC1, the term "Prepared" demands an escrowed payment, which is not accurate in an optimistic-mode transfer.  However, if we use the term `created_at`, but in some instances (e.g., Universal Transfer) we really mean "when this was prepared", it it could lead to confusion.  Likewise, choosing the value `prepared_at` when  we really mean "when this was created" (e.g., Optimistic Transfer), seems equally imprecise.

Along those lines, RFC1 talks about a `Proposed` state.  Would it make sense to allow transfers to be created in the `Proposed` state initially, and then (potentially at some point later) moved into the `Prepared` state, but only if its a conditional transfer?  

That would allow optimistic transfers to have a timeline like the following:

```
"timeline": {
  "created_at": "2015-06-16T00:00:01.000Z",
  "executed_at": "2015-06-16T00:00:02.000Z"
}
```

whereas conditional transfers could have a timeline like this:

```
"timeline": {
  "created_at": "2015-06-16T00:00:01.000Z",
  "prepared_at": "2015-06-16T00:00:01.000Z",
  "executed_at": "2015-06-16T00:00:02.000Z"
}
```